oak_functions/loader/src/metrics.rs

```diff
@@ -0,0 +1,169 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::logger::Logger;
+use log::Level;
+use rand::{distributions::Open01, thread_rng, Rng};
+use serde::Deserialize;
+use std::{
+    collections::{HashMap, HashSet},
+    sync::{Arc, Mutex},
+};
+
+/// Configuration for differentially-private metrics reporting.
+#[derive(Deserialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct PrivateMetricsConfig {
+    pub epsilon: f64,
+    pub batch_size: usize,
+    pub allowed_labels: Vec<String>,
+}
+
+/// Aggregator for count-based differentially private metrics.
+pub struct PrivateMetricsAggregator {
+    count: usize,
+    epsilon: f64,
+    batch_size: usize,
+    allowed_labels: HashSet<String>,
+    label_count: f64,
+    events: HashMap<String, usize>,
+    logger: Logger,
+}
+
+impl PrivateMetricsAggregator {
+    pub fn new(config: &PrivateMetricsConfig, logger: Logger) -> Self {
+        Self {
+            count: 0,
+            epsilon: config.epsilon,
+            batch_size: config.batch_size,
+            allowed_labels: config.allowed_labels.iter().cloned().collect(),
+            label_count: config.allowed_labels.iter().count() as f64,
+            events: HashMap::new(),
+            logger,
+        }
+    }
+
+    /// Reports new events that should be included in the aggregated counts.
+    pub fn report_events(&mut self, events: HashSet<String>) {
+        self.count += 1;
+        for label in self.allowed_labels.intersection(&events) {
+            if let Some(bucket) = self.events.get_mut(label) {
+                *bucket += 1;
+            } else {
+                self.events.insert(label.to_owned(), 1);
+            }
+        }
+
+        if self.count == self.batch_size {
+            self.export_events();
+        }
+    }
+
+    /// Logs the current counts after adding appropriate noise, resets the batch count to 0 and
+    /// clears the aggregated data.
+    fn export_events(&mut self) {
+        let counts: Vec<String> = self
+            .allowed_labels
+            .iter()
+            .map(|label| {
+                let value = self.add_laplace_noise(*self.events.get(label).unwrap_or(&0));
+                format!("{}={}", label, value)
+            })
+            .collect();
+        self.logger.log_public(
+            Level::Info,
+            &format!(
+                "metrics export: batch size: {}; counts: {}",
+                self.count,
+                counts.join(";"),
+            ),
+        );
+        self.count = 0;
+        self.events.clear();
+    }
+
+    /// Exports the batch count and bucket counts without adding any noise for use in testing. Also
+    /// resets the batch count to 0 and clears the aggregated data.
+    pub fn export_events_for_test(&mut self) -> (usize, Vec<(String, usize)>) {
```

Comment:
Both these methods are called `export_*`, but one logs, while the other returns the things to log. Maybe rename them, or make them more consistent? e.g. this method may log non-publicly, which would work only in the unsafe version anyways.